text/0000-inferred-type-aliases.md

```diff
@@ -0,0 +1,696 @@
+unnameable types are the types of closures. Each closure has a distinct type,
+which the compiler knows about, but you can't write their name in code.
+
+An example:
+
+```rust
+#![feature(core_intrinsics)]
+
+use std::intrinsics::type_name;
+
+fn name_of_type<T>(_: &T) -> &str {
+    unsafe { type_name::<T>() }
+}
+
+fn main() {
+    let closure_one = |x: usize| x + 1;
+    let closure_two = |x: usize| x + 1;
+    println!("{}", name_of_type(&closure_one));
+    // output: [closure@src/main.rs:10:23: 10:39]
+    println!("{}", name_of_type(&closure_two));
+    // output: [closure@src/main.rs:11:23: 11:39]
+}
+```
+
+As you can see from the output, the type of these two closures are *not* the same.
+
+## The Proposal
+[the proposal]: #the-proposal
+
+Now that we've defined the core concepts this RFC deals with,
+let's deal with what is actually proposed.
+
+This part is rather simple.
+What this RFC proposes is that you should be able to have the actual type of a
+`type` alias as well as the associated type of an implementation be [inferred]
+and that these inferred types should be [transparent] as opposed to [opaque].
+We propose to do this by using the familiar symbol `_` which means
+*"please infer the type"* in type contexts.
+
+For example, you will be allowed to write:
+
+```rust
+use std::ops::Range;
+
+type MyIterator = _;
+
+fn one_to_hundred() -> MyIterator {
+    let range: Range<u8> = 0..100;
+    range
+}
+
+fn main() {
+    // OK! because this is transparent inference.
+    let _numbers: Range<u8> = one_to_hundred();
+}
+```
+
+You can also optionally enforce that `MyIterator` must at least be an `Iterator`:
+
+```rust
+type MyIterator: Iterator<Item = u8> = _;
+```
+
+This works the same way in associated types:
+
+```rust
+// Let's pretend this type is more interesting in actuality.
+#[derive(Debug)]
+struct SomeImportantStuff;
+struct MyType;
+trait MyTrait {
+    type MyImplementationDetail;
+    fn my_public_api(&self) -> Self::MyImplementationDetail;    
+}
+
+// Our implementation:
+impl MyTrait for MyType {
+    // Here we employ type inference to our aid:
+    type MyImplementationDetail = _;
+
+    fn my_public_api(&self) -> Self::MyImplementationDetail {
+        // Here goes logic that determines the
+        // actual type of `MyImplementationDetail`...
+        SomeImportantStuff
+    }
+}
+
+fn main() {
+    // OK! Inferred type is transparent.
+    let _detail: SomeImportantStuff = MyType.my_public_api();
+}
+```
+
+Same as with type aliases, you may also enforce a trait bound:
+
+```rust
+    // Still transparent and will leak everything about the type:
+    type MyImplementationDetail: Debug = _;
+```
+
+Another way to think of the feature proposed in this RFC is that it is the
+`_` inside `type Alias = _;` is the [transparent] version of the [opaque]
+and [inferred] `type Alias = impl Trait` ([RFC 2515]) or equivalently
+`existential type Alias: Trait;` ([RFC 2071]).
```

Comment:
It would be good to discuss the exact relationship between the two in terms of whether one is representable by the other. If `existential type` can be explained using a desugaring into `_`, that would be very helpful.

Reply:
For example, something like:
```rust
type Foo = impl Bar;
// desugars into:
struct {Foo}(_: Bar);
type Foo = {Foo};
impl Bar for Foo {
	// delegate everything to self.0
}
```
You have to explain the inference differences, though (e.g. not inferring across function boundaries for TAIT), but I think that's doable.